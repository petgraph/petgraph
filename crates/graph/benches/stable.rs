mod common;

use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion};
use petgraph_core::{edge::Direction, visit::IntoEdges};
use petgraph_graph::NodeIndex;

use crate::common::{cycle_stable_graph, nodes};

fn edges_count(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("edges/count");

    for size in nodes(None) {
        group.bench_with_input(BenchmarkId::from_parameter(size), size, |bench, &size| {
            bench.iter_batched(
                || cycle_stable_graph::<(), ()>(size, None),
                |(graph, ..)| {
                    let _count = graph.edges(NodeIndex::new(0)).count();
                },
                BatchSize::SmallInput,
            );
        });
    }
}

fn edges_directed_count(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("edges_directed/count");

    for size in nodes(None) {
        group.bench_with_input(BenchmarkId::new("Outgoing", size), size, |bench, &size| {
            bench.iter_batched(
                || cycle_stable_graph::<(), ()>(size, None),
                |(graph, ..)| {
                    let _count = graph
                        .edges_directed(NodeIndex::new(0), Direction::Outgoing)
                        .count();
                },
                BatchSize::SmallInput,
            );
        });

        group.bench_with_input(BenchmarkId::new("Incoming", size), size, |bench, &size| {
            bench.iter_batched(
                || cycle_stable_graph::<(), ()>(size, None),
                |(graph, ..)| {
                    let _count = graph
                        .edges_directed(NodeIndex::new(0), Direction::Outgoing)
                        .count();
                },
                BatchSize::SmallInput,
            );
        });
    }
}

fn neighbours_count(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("neighbors/count");

    for size in nodes(None) {
        group.bench_with_input(BenchmarkId::from_parameter(size), size, |bench, &size| {
            bench.iter_batched(
                || cycle_stable_graph::<(), ()>(size, None),
                |(graph, ..)| {
                    let _count = graph.neighbors(NodeIndex::new(0)).count();
                },
                BatchSize::SmallInput,
            );
        });
    }
}

fn neighbours_directed_count(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("neighbors_directed/count");

    for size in nodes(None) {
        group.bench_with_input(BenchmarkId::new("Outgoing", size), size, |bench, &size| {
            bench.iter_batched(
                || cycle_stable_graph::<(), ()>(size, None),
                |(graph, ..)| {
                    let _count = graph
                        .neighbors_directed(NodeIndex::new(0), Direction::Outgoing)
                        .count();
                },
                BatchSize::SmallInput,
            );
        });

        group.bench_with_input(BenchmarkId::new("Incoming", size), size, |bench, &size| {
            bench.iter_batched(
                || cycle_stable_graph::<(), ()>(size, None),
                |(graph, ..)| {
                    let _count = graph
                        .neighbors_directed(NodeIndex::new(0), Direction::Incoming)
                        .count();
                },
                BatchSize::SmallInput,
            );
        });
    }
}

fn map_(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("map");

    for size in nodes(None) {
        group.bench_with_input(BenchmarkId::from_parameter(size), size, |bench, &size| {
            bench.iter_batched(
                || cycle_stable_graph::<u8, u8>(size, None),
                |(graph, ..)| {
                    let _mapped = graph.map(
                        |_, weight| weight.saturating_mul(2),
                        |_, weight| weight.saturating_mul(3),
                    );
                },
                BatchSize::SmallInput,
            );
        });
    }
}

fn retain_nodes(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("retain/nodes");

    for size in nodes(None) {
        group.bench_with_input(BenchmarkId::from_parameter(size), size, |bench, &size| {
            bench.iter_batched(
                || {
                    let (graph, ..) = cycle_stable_graph::<(), ()>(size, None);
                    let remove = graph.node_count() / 2;

                    (graph, remove)
                },
                |(mut graph, remove)| {
                    graph.retain_nodes(|_, index| index.index() > remove);
                },
                BatchSize::SmallInput,
            );
        });
    }
}

fn retain_edges(criterion: &mut Criterion) {
    let mut group = criterion.benchmark_group("retain/edges");

    for size in nodes(None) {
        group.bench_with_input(BenchmarkId::from_parameter(size), size, |bench, &size| {
            bench.iter_batched(
                || {
                    let (graph, ..) = cycle_stable_graph::<(), ()>(size, None);
                    let remove = graph.edge_count() / 2;

                    (graph, remove)
                },
                |(mut graph, remove)| {
                    graph.retain_edges(|_, index| index.index() > remove);
                },
                BatchSize::SmallInput,
            );
        });
    }
}

criterion_group!(edges, edges_count, edges_directed_count);
criterion_group!(neighbours, neighbours_count, neighbours_directed_count);
criterion_group!(retain, retain_nodes, retain_edges);
criterion_group!(map, map_);

criterion_main!(edges, neighbours, retain, map);
