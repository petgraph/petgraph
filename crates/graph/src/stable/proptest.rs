use core::fmt::Debug;
use std::{collections::BTreeSet, sync::Arc};

use petgraph_core::{edge::EdgeType, index::IndexType};
use petgraph_proptest::default::graph_strategy;
use proptest::{
    arbitrary::Arbitrary,
    bits::BitSetLike,
    collection::{btree_set, SizeRange},
    prelude::{BoxedStrategy, Just, Strategy},
    strategy::{LazyJust, TupleUnion},
};

use crate::{stable::StableGraph, NodeIndex};

/// `Arbitrary` for `Graph` creates a graph by selecting a node count
///
/// The result will be simple graph or digraph, self loops
/// possible, no parallel edges.
///
/// The exact properties of the produced graph is subject to change.
///
/// Requires crate feature `"proptest"`
impl<N, E, Ty, Ix> Arbitrary for StableGraph<N, E, Ty, Ix>
where
    N: Arbitrary + Debug + Clone + 'static,
    E: Arbitrary + Debug + Clone + 'static,
    Ty: EdgeType + Send + 'static,
    Ix: IndexType + Send,
{
    type Parameters = ();
    // impl Strategy<Value = Self> is nightly, and therefore not usable here.
    // TODO: revisit once impl_trait_in_assoc_type is stable. (https://github.com/rust-lang/rust/issues/63063)
    type Strategy = BoxedStrategy<Self>;

    fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
        // `MAX` is not a valid node index.
        graph_strategy(
            true,
            false,
            0..Ix::MAX.as_usize(),
            Some(Arc::new(|max| {
                SizeRange::new(0..=usize::min(max.pow(2), Ix::MAX.as_usize() - 1))
            })),
        )
        .prop_flat_map(|graph: Self| {
            let nodes = graph.node_count();
            let is_empty = nodes == 0;

            let removed_nodes = TupleUnion::new((
                (u32::from(is_empty), Arc::new(LazyJust::new(BTreeSet::new))),
                (
                    u32::from(!is_empty),
                    Arc::new(btree_set(0..nodes.len(), 0..nodes.len())),
                ),
            ));

            // select a batch of random indices to remove (unique)
            (Just(graph), removed_nodes)
        })
        .prop_map(|(mut graph, remove)| {
            for index in remove {
                graph.remove_node(NodeIndex(Ix::from_usize(index)));
            }

            graph
        })
        .boxed()
    }
}
