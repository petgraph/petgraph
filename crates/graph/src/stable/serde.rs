use alloc::vec::Vec;
use core::{cell::Cell, marker::PhantomData};

use funty::Integral;
use petgraph_core::deprecated::{
    index::IndexType,
    visit::{EdgeIndexable, NodeIndexable},
};
use serde::{
    de::Error as _,
    ser::{SerializeSeq, SerializeStruct},
    Deserialize, Deserializer, Serialize, Serializer,
};

use crate::{
    serde::{EdgeProperty, InvalidError},
    stable::StableGraph,
    Edge, EdgeIndex, EdgeType, Graph, Node, NodeIndex,
};

struct SerializeIter<T> {
    len: usize,
    iter: Cell<Option<T>>,
}

impl<T> SerializeIter<T> {
    const fn new(len: usize, iter: T) -> Self {
        Self {
            len,
            iter: Cell::new(Some(iter)),
        }
    }
}

impl<T> Serialize for SerializeIter<T>
where
    T: Iterator,
    T::Item: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let iter = self.iter.take().expect("SerializeIter already serialized");

        let mut seq = serializer.serialize_seq(Some(self.len))?;
        let mut count = 0;
        for item in iter {
            seq.serialize_element(&item)?;
            count += 1;
        }

        debug_assert_eq!(count, self.len, "SerializeIter length mismatch");
        seq.end()
    }
}

impl<N, E, Ty, Ix> Serialize for StableGraph<N, E, Ty, Ix>
where
    N: Serialize,
    E: Serialize,
    Ty: EdgeType,
    Ix: Serialize + IndexType,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let nodes = &self.raw_nodes()[..self.node_bound()];
        let node_count = self.node_count();
        let hole_count = nodes.len() - node_count;
        let edges = &self.raw_edges()[..self.edge_bound()];

        let mut struct_ = serializer.serialize_struct("StableGraph", 4)?;

        struct_.serialize_field(
            "nodes",
            &SerializeIter::new(
                node_count,
                nodes.iter().filter_map(|node| node.weight.as_ref()),
            ),
        )?;
        struct_.serialize_field(
            "node_holes",
            &SerializeIter::new(
                hole_count,
                nodes.iter().enumerate().filter_map(|(index, node)| {
                    node.weight.is_none().then(|| NodeIndex::<Ix>::new(index))
                }),
            ),
        )?;
        struct_.serialize_field("edge_property", &EdgeProperty::from_type::<Ty>())?;

        // Convert from `Edge<Option<E, Ix>>` to `Option<Edge<E, Ix>>`
        let edges_len = edges.len();
        struct_.serialize_field(
            "edges",
            &SerializeIter::new(
                edges_len,
                edges.iter().map(|edge| {
                    edge.weight.as_ref().map(|weight| Edge {
                        weight,
                        next: edge.next,
                        node: edge.node,
                    })
                }),
            ),
        )?;

        struct_.end()
    }
}

#[derive(Deserialize)]
#[serde(rename = "Graph")]
struct RemoteStableGraph<N, E, Ix: IndexType> {
    nodes: Vec<Node<N, Ix>>,
    #[serde(default)]
    node_holes: Vec<NodeIndex<Ix>>,
    edge_property: EdgeProperty,
    edges: Vec<Option<Edge<E, Ix>>>,
}

impl<'de, N, E, Ty, Ix> Deserialize<'de> for StableGraph<N, E, Ty, Ix>
where
    N: Deserialize<'de>,
    E: Deserialize<'de>,
    Ty: EdgeType,
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let RemoteStableGraph {
            nodes,
            node_holes,
            edge_property,
            edges,
        } = RemoteStableGraph::<N, E, Ix>::deserialize(deserializer)?;

        // convert from `Option<Edge<E, Ix>>` to `Edge<Option<E>, Ix>`
        let edges = edges
            .into_iter()
            .map(|edge| match edge {
                Some(Edge { weight, next, node }) => Edge {
                    weight: Some(weight),
                    next,
                    node,
                },
                None => Edge {
                    weight: None,
                    next: [EdgeIndex::end(); 2],
                    node: [NodeIndex::end(); 2],
                },
            })
            .collect::<Vec<_>>();

        if edges.len() >= <Ix as Integral>::MAX.as_usize() {
            return Err(InvalidError::Length {
                type_: "edge",
                length: edges.len(),
                max: Ix::MAX.as_usize(),
            })
            .map_err(D::Error::custom);
        }

        let expected = EdgeProperty::from_type::<Ty>();
        if edge_property != expected {
            return Err(InvalidError::Direction {
                expected,
                received: edge_property,
            })
            .map_err(D::Error::custom);
        }

        let total_nodes = nodes.len() + node_holes.len();
        let mut compact_nodes = nodes.into_iter().map(|node| Node {
            weight: Some(node.weight),
            next: node.next,
        });

        let mut nodes = Vec::with_capacity(total_nodes);

        let mut node_pos = 0;
        for position in &node_holes {
            let position = position.index();

            if !(node_pos..total_nodes).contains(&position) {
                return Err(InvalidError::Hole { index: position }).map_err(D::Error::custom);
            }

            nodes.extend(compact_nodes.by_ref().take(position - node_pos));
            nodes.push(Node {
                weight: None,
                next: [EdgeIndex::end(); 2],
            });

            node_pos = position + 1;
            debug_assert_eq!(nodes.len(), node_pos);
        }

        nodes.extend(compact_nodes);

        if nodes.len() >= <Ix as Integral>::MAX.as_usize() {
            return Err(InvalidError::Length {
                type_: "node",
                length: nodes.len(),
                max: Ix::MAX.as_usize(),
            })
            .map_err(D::Error::custom);
        }

        let node_bound = nodes.len();
        let mut stable_graph = Self {
            g: Graph {
                nodes,
                edges,
                ty: PhantomData,
            },
            node_count: 0,
            edge_count: 0,
            free_edge: EdgeIndex::end(),
            free_node: NodeIndex::end(),
        };

        stable_graph
            .link_edges()
            .map_err(|i| InvalidError::Node {
                index: i.index(),
                length: node_bound,
            })
            .map_err(D::Error::custom)?;

        Ok(stable_graph)
    }
}

#[cfg(test)]
mod tests {
    use alloc::vec::Vec;

    use serde::Deserialize;

    use crate::{
        node_index,
        serde::EdgeProperty,
        stable::{serde::RemoteStableGraph, StableUnGraph},
        EdgeIndex, Node,
    };

    #[test]
    fn deserialization_with_holes() {
        let mut graph = StableUnGraph::<i32, ()>::with_capacity(0, 0);

        let a = graph.add_node(0);
        let b = graph.add_node(1);
        let c = graph.add_node(2);
        let d = graph.add_node(3);
        let e = graph.add_node(4);
        let f = graph.add_node(5);
        let g = graph.add_node(6);

        graph.remove_node(a);
        graph.remove_node(c);
        graph.remove_node(d);
        // as optimization technique, we use `node_bound()` for raw_nodes() and raw_edges(), this
        // means that if we have a trailing hole it will be skipped.
        graph.remove_node(g);

        let value = serde_value::to_value(&graph).unwrap();
        let graph = StableUnGraph::<i32, ()>::deserialize(value).unwrap();

        assert_eq!(graph.node_count(), 3);
        assert_eq!(
            graph
                .raw_nodes()
                .iter()
                .map(|n| n.weight)
                .collect::<Vec<_>>(),
            vec![None, Some(1), None, None, Some(4), Some(5)],
        );
    }
}
