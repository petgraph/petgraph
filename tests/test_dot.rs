#![cfg(feature = "dot")]

use insta::{assert_debug_snapshot, assert_display_snapshot};
#[cfg(feature = "stable-graph")]
use petgraph::graph::stable::StableGraph;
use petgraph::{dot::Dot, Graph};
#[cfg(feature = "adjacency-matrix")]
use petgraph_adjacency_matrix::AdjacencyList;

#[test]
#[cfg(feature = "adjacency-matrix")]
fn adjacency_list() {
    let mut graph = AdjacencyList::new();

    let a = graph.add_node();
    let b = graph.add_node();

    graph.add_edge(a, a, "A -> A");
    graph.add_edge(a, b, "A -> B");
    graph.add_edge(b, a, "B -> A");

    assert_debug_snapshot!(Dot::new(&graph));
}

#[test]
fn graph() {
    let mut graph = Graph::new();

    let a = graph.add_node("A");
    let b = graph.add_node("B");

    graph.add_edge(a, a, "A -> A");
    graph.add_edge(a, b, "A -> B");
    graph.add_edge(b, a, "B -> A");

    assert_debug_snapshot!(Dot::new(&graph));
}

#[test]
fn debug_formatting() {
    // test alternate formatting
    #[derive(Debug)]
    struct Record {
        a: i32,
        b: &'static str,
    }
    let mut graph = Graph::new();
    let a = graph.add_node(Record { a: 1, b: r"abc\" });
    graph.add_edge(a, a, (1, 2));

    assert_debug_snapshot!(Dot::new(&graph));
}

#[test]
#[cfg(feature = "stable-graph")]
fn stable_graph() {
    let mut graph = StableGraph::new();

    let a = graph.add_node("A");
    let b = graph.add_node("B");

    graph.add_edge(a, a, "A -> A");
    graph.add_edge(a, b, "A -> B");

    assert_debug_snapshot!(Dot::new(&graph));
}
