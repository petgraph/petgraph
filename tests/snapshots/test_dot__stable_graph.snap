---
source: tests/test_dot.rs
expression: "Dot::new(&graph)"
---
digraph petgraph {
    N0[label=""];
    N1[label=""];
    N0 -> N0[label=""];
    N0 -> N1[label=""];
}

